storage: factor out node dialer from RaftTransport

tbg opened the pull request:
RaftTransport wraps RPC connections between nodes with additional
circuit breaker logic to avoid spamming dead nodes (plus spamming the
logs). Upcoming work on follower reads needs similar machinery, and
there are various other components in the system that connect to other
nodes and should already be using such a mechanism.

This commit extracts this logic into a standalone component named
`NodeDialer`. In future work, a single `NodeDialer` will be shared
across a running node.

There is a semantic change in behavior that deserves being called out.
Prior to this change, the Raft circuit breakers were tripped not only
based on whether it was possible to connect to a given target node, but
also based on error codes returned from the calls, and notably
StoreNotFoundError.
This is no longer the case, as it would be at odds with the goal of
sharing these circuit breakers across multiple use cases and maintaining
the old behavior would require adding another layer of circuit breaking
inside the Raft transport.

Looking at the history of this behavior (#10266), it seems to have been
introduced because bringing a new node up under the address of an old
node could wedge the (ancient version of the) transport indefinitely as
it didn't reconnect in that case (taking the correct address from
Gossip) but kept sending messages to the wrong node. This is still fixed
with my changes; only the backoff is lost, which seems acceptable.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/26977)
<!-- Reviewable:end -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 4 files at r1.
Review status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

---



*Comments from [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/26977#-:-LFxD8ugBrTjUun-LAIY:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: factor out node dialer from RaftTransport

RaftTransport wraps RPC connections between nodes with additional
circuit breaker logic to avoid spamming dead nodes (plus spamming the
logs). Upcoming work on follower reads needs similar machinery, and
there are various other components in the system that connect to other
nodes and should already be using such a mechanism.

This commit extracts this logic into a standalone component named
`NodeDialer`. In future work, a single `NodeDialer` will be shared
across a running node.

There is a semantic change in behavior that deserves being called out.
Prior to this change, the Raft circuit breakers were tripped not only
based on whether it was possible to connect to a given target node, but
also based on error codes returned from the calls, and notably
StoreNotFoundError.
This is no longer the case, as it would be at odds with the goal of
sharing these circuit breakers across multiple use cases and maintaining
the old behavior would require adding another layer of circuit breaking
inside the Raft transport.

Looking at the history of this behavior (#10266), it seems to have been
introduced because bringing a new node up under the address of an old
node could wedge the (ancient version of the) transport indefinitely as
it didn't reconnect in that case (taking the correct address from
Gossip) but kept sending messages to the wrong node. This is still fixed
with my changes; only the backoff is lost, which seems acceptable.

Release note: None

tschottdorf commented:
TFTR!

bors r=bdarnell
